Introduction to Hooks (#1545)
enkidevs/curriculum: javascript/react/introduction-to-hooks/effect-hooks.md

---

# Effect hooks

---
## Content

Effect hooks allow you to perform side effects in functional components. Some examples of side effects include setting up a subscription, manually changing the DOM or data fetching. Let's use the same `<Enki />` component that was created in the previous insight:

```jsx
import React, {
  useState,
  useEffect
} from "react";

function Enki() {
  const [name, setName] = useState("Enki");

  useEffect(() => {
    document.title = `Welcome ${name}.`;
  });

  return <p>Hello, my name is {name}</p>;
}
```

Notice anything new? We have used the `useEffect()` hook to set the document's title to a new custom message that includes the `name` state. If the syntax looks a bit different it is because we are using template literals[1] that allow us to embed expressions. You can think of the `useEffect()` hook as a replacement for `componentDidMount`, `componentDidUpdate`, and `componentWillUnmount` combined.

Note that you have to import the `useEffect` hook from the `'react'` library.

There are two types of effect hooks: with cleanup or without cleanup. In this insight we will focus on the effects without cleanup.

> When using class components, the side effects are defined inside the `componentDidMount`, `componentDidUpdate` or `componentWillUnmount` methods, because defining them in the `render` method would be too early (we want side effects to happen **after** React updates the DOM). Here is how our example would look like if we used class components:

Reviewer: What do you think of?

```suggestion
When using class components, a common pattern for side effects is to define them in the `componentDidMount` method, update them in the `componentDidUpdate` method, and clean them up in the `componentWillUnmount` method.

Here is how our example would look like if we used class components:
```

Note: I removed one sentence because we probably don't always want to have side effects after render

```jsx
class Enki extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      name: "Enki"
    };
  }

  componentDidMount() {
    document.title = `Welcome ${name}.`;
  }

  componentDidUpdate() {
    document.title = `Welcome ${name}.`;
  }

  render() {
    return (
      <p>Hello, my name is {this.name}.</p>
    );
  }
}
```

> You can already see the advantages of using hooks by comparing these two examples. With the `useEffect()` hook, you tell React that your component has to do something after every render. The `useEffect()` hook is placed inside the component so that we can access the `name` variable. This is possible because of how JavaScript's scope works, and it removes the need of React-specific APIs that would have to read the hook. By default, the `useEffect()` is ran after the first render and after every subsequent update.

Reviewer: I assume sometime later we'll explain how to control when effects run by using their dependencies?

```suggestion
You can already see the advantages of using hooks by comparing these two examples. With the `useEffect()` hook, you tell React that your component has to do something after every render. The `useEffect()` hook is placed inside the component so that we can access the `name` variable. This is possible because of how JavaScript's scope works, and it removes the need of React-specific APIs that would have to read the hook. By default, the `useEffect()` is ran after the first render and after every subsequent update (we'll see more how to control this later on).
```

---
## Practice

The `useEffect()` hook can be considered as a replacement of what when using functional components?

???

* All of the methods mentioned.
* `componentDidUpdate()`
* `componentDidMount()`
* `componentWillUnmount()`

---
## Revision

> Complete the following code to change the document's title to "Welcome to the revision question.":

Reviewer: ```suggestion
Complete the following code to change the document's title to `"Welcome to the revision question."`:
```
